### ISSUE TITLE
Idea: `@psalm-never-throw`, alike to `@psalm-pure`, but promises to not throw/warn

### ISSUE THREAD

**Issue by Ocramius** [vimeo/psalm#2912]
Similar to `@psalm-pure`, it may be **interesting** to have something like `@psalm-never-throw`, in which the guarantees are:

 1. the code does not kill the process (yes, the engine can still go out of memory or out of stack frames)
 2. the code does not raise any notices/warnings
 3. the code does not throw exceptions
 4. all thrown exceptions are caught

This is interesting when declaring interfaces with strong guarantees, such as:

```php
/** @psalm-never-throw */
interface Worker
{
    function process(Queue $queue): void;
}
```

The code in any `Worker#run()` implementations must therefore only rely on other `@psalm-never-throw` implementations, or catch any exceptions declared in code that has `@throws`.

Triggering errors/warnings, as well as the `throw` construct are forbidden in classes/functions annotated with `@psalm-never-throw`.

No further guarantees are really possible with the language, but this should play nicely with libraries such as `webmozart/assert`, or `thecodingmachine/safe`.

Again, feel free to shoot this down: I'm only *throwing* this out here as an idea :wink:

**Comment by letnando** [comment]
Good idea, the naming you proposed is more verbose but I would propose `@psalm-safe`, meaning that's safe to be called without throwing exceptions/errors. What do you think? 🤔

**Comment by Ocramius** [comment]
@letnando `@psalm-safe` seems easily confused with `@psalm-pure`. For me, a method that does not throw, yet does modify global state, is not safe :woman_shrugging:

**Comment by bdsl** [comment]
I wonder if `@psalm-throw-only` would be a good addition to this, to guarantee that a function only throws exceptions that it declares with `@throws`. 

Then presumably a `@psalm-never-throw` function could call an `@psalm-throw-only` function as long as it catches all thrown exceptions.

It's a bit reminiscent of Java's checked exceptions, but with function by function opt-in checking.

**Comment by Ocramius** [comment]
@bdsl doesn't `@psalm-never-return` already do that, when combined with `@throws`?

**Comment by bdsl** [comment]
Ah no, I didn't mean that a `psalm-throw-only` function must not return - I meant that it must not throw or exit *except* by throwing one of its declared exceptions. It can still return.

**Comment by M1ke** [comment]
For that case would it just make sense to treat `@throws` as that guarantee anyway? Like, if you're saying `@throws FileNotFoundException` but your code can also throw `IllegalArgumentException` you're being misleading surely?

**Comment by ondrejmirtes** [comment]
PHPStan uses `@throws void` for this. For example if the parent method has some `@throws Exception` and you don't want to inherit these tags, you need to use `@throws void` above the overriden method.

**Comment by ondrejmirtes** [comment]
One package that I think takes advantage of this is https://github.com/pepakriz/phpstan-exception-rules/.

**Comment by Ocramius** [comment]
Didn't know about `@throws void`: interesting solution that doesn't need additional/new syntax!

**Comment by sebastianbergmann** [comment]
I would prefer `@psalm-never-throw` over `@throws void`.

**Comment by bdsl** [comment]
If we wanted a solution without new syntax I'd write it as `@throws never` rather than thows void. `never` already exists in Psalm as an alias for the bottom type `never-return`.

**Comment by bdsl** [comment]
@M1ke I don't think I'd expect `@throws` to be a gurantee the nothing else will be thrown - I only use `@throws` to declare exceptions that I think the imediate caller might be interested in catching or redeclaring. For instance I wouldn't generally declare `@throws Doctrine\DBAL\Exception` on a repository.

**Comment by ondrejmirtes** [comment]
@bdsl Yes, this distinction is called checked/unchecked exceptions. You should declare `throws` and be catching/rethrowing exceptions that are interesting for the application logic and are expected to happen. Something like `CustomerCanNoLongerBuyThisProductException`.

Unchecked exceptions are for stuff that can break but shouldn't be handled in code, but instead prevented by validations, or fixed in infrastructure. This for example includes `PDOException` (when the database disconnects) or `NegativeIntegerException`.

**Comment by muglug** [comment]
I have a couple of thoughts, but I'm not sure of their usefulness.

1. We could make throwing inside a pure/mutation-free context (or calling a function/method that can throw) a specific sort of issue (e.g. `ThrowInsidePure`)
2. The code that triggers `ThrowInsidePure` would be subject to a config flag being manually set to `false`, maybe `allowExceptionsInsidePureFunctions` or similar.

This would mean that a given project's maintainer could decide to opt in to this behaviour globally, then suppress it either at the directory, file, class or function level.

I do something similar for `throws` checks – if `checkForThrowsDocblock="true"` in your config then Psalm will emit `MissingThrowsDocblock` issues.

**Comment by Ocramius** [comment]
Not really what I had in mind, since the example above (workers, listeners, etc) are all but pure code 🤔

**Comment by muglug** [comment]
Oh sorry, I thought this was an extension to pure functions & immutable classes, not a separate (but related) track

**Comment by muglug** [comment]
In that case this really is identical to  `@psalm-immutable`/`@psalm-pure` in terms of enforcement – if I'm reading it correctly the property must be on every single call:

```php
/**
 * @psalm-never-throws
 */
class SomeClass {
  public function someMethod(Foo $foo) {
    $foo->thisMethodMustHaveANeverThrowAnnotation();
    echo (string) $foo; // so should this  __toString()
    echo $foo->aMagicGetWouldNeedItToo;
  }
}
```

I can see the actual _downstream_ enforcement of that being a bit of a slog – you'd have to make a potentially non-trivial amount of code changes to pass these checks (the same is obviously true of pure/immutable annotations, but at least that cost/benefit ratio is widely understood by the FP community).

This is not something I'm likely to implement want to implement in Psalm for myself, but I'll gladly accept a PR – I just created a special `mattwontfix` label. For this and other similar things where I invite others' contributions!
